web/src/users/login.rs

```diff
@@ -0,0 +1,110 @@
+use axum::{
+    extract::State,
+    response::{IntoResponse, Redirect},
+    Form,
+};
+use cja::{
+    app_state::AppState as _,
+    tower_cookies::{Cookie, Cookies},
+};
+
+use db::users::User;
+
+use password_auth::verify_password;
+
+use crate::{sessions::Session, templates::Template, AppState, WebResult};
+
+pub async fn get(t: Template) -> impl IntoResponse {
+    t.render(maud::html! {
+        form method="post" action="/login" {
+            input type="text" name="username" placeholder="Username";
+            input type="password" name="password" placeholder="Password";
+            input type="submit" value="Login";
+        }
+    })
+}
+
+#[derive(serde::Deserialize)]
+pub struct FromData {
+    username: String,
+    password: String,
+}
+
+#[axum::debug_handler(state = AppState)]
+pub async fn post(
```

Comment:
And this is what gets called when you submit that form!

We get the `FormData` (typoed as FromData here lol) by telling Axum the user filled out a form and we want the data to fit into our `FromData` struct!

We also need the `Session` and the `AppState` so we can record the user as logged in!


We first look up the user they asked for by username, then we verify if the password is correct. If it is NOT we redirect back to login with a flash message.
If the password is valid we save the `user_id` to the `Session` which logs the user in!